tv/tv.cpp

```diff
@@ -591,56 +605,147 @@ bool do_skip(const llvm::StringRef &pass0) {
     [](llvm::PassBuilder &PB) {
       is_clangtv = true;
+      PB.registerPipelineParsingCallback(
+          [](llvm::StringRef Name,
+             llvm::ModulePassManager &MPM,
+             llvm::ArrayRef<llvm::PassBuilder::PipelineElement>) {
+          static bool first_tv = true;
+          if (Name != "tv")
+            return false;
+
+          is_from_npm = true;
+          if (first_tv) {
+            // Assume that this plugin is loaded from opt when tv pass is
+            // explicitly given as an argument
+            is_clangtv = false;
+            first_tv = false;
+            MPM.addPass(TVInitPass());
+            MPM.addPass(TVNewPass());
+          } else {
+            MPM.addPass(TVNewPass());
+            // Finalization is done by the last TVNewPass
+          }
+          TVNewPass::num_tv_pass_created++;
+          return true;
+        });
+      // These two register*EPCallbacks are called when 'default' pipelines
+      // such as O2, O3 are used.
       PB.registerPipelineStartEPCallback(
           [](llvm::ModulePassManager &MPM,
              llvm::PassBuilder::OptimizationLevel) {
+            is_from_npm = true;
             MPM.addPass(TVInitPass());
           });
       PB.registerOptimizerLastEPCallback(
           [](llvm::ModulePassManager &MPM,
              llvm::PassBuilder::OptimizationLevel) {
-            MPM.addPass(createModuleToFunctionPassAdaptor(TVFinalizePass()));
+            MPM.addPass(TVFinalizePass());
           });
       auto f = [](llvm::StringRef P, llvm::Any IR,
                   const llvm::PreservedAnalyses &PA) {
-        static int count = 0;
-        if (!out) {
-          // TVInitPass is not called yet.
-          // This can happen at very early passes, such as
-          // ForceFunctionAttrsPass.
+        if (TVFinalizePass::finalized)
           return;
-        }
 
-        if (TVFinalizePass::finalized)
+        TVNewPass::pass_name = P.str();
+        TVNewPass::skip_tv |= do_skip(TVNewPass::pass_name);
+
+        if (!is_clangtv)
           return;
 
-        bool skip_pass = do_skip(P);
-        *out << "-- " << ++count << ". " << P.str()
-             << (skip_pass ? " : Skipping\n" : "\n");
-
-        TVPass tv;
-        tv.skip_verify = skip_pass;
-        // For I/O known calls like printf, it is fine to regard them as valid
-        // 'unknown calls' except when it is InstCombine.
-        tv.encode_io_fns_as_unknown = P != "InstCombinePass" &&
-                                      P != "AggressiveInstCombinePass";
-
-        auto M = const_cast<llvm::Module *>(unwrapModule(IR));
-        for (auto &F: *M)
-          // If skip_pass is true, this updates fns map only.
-          tv.runOnFunction(F);
+        // If it is clang tv, validate each pass
+        TVNewPass tv;
+        tv.print_pass_name = true;
+
+        llvm::TargetLibraryInfoImpl TLIImpl;
+        unique_ptr<llvm::TargetLibraryInfo> TLI_holder;
```

Comment:
you can set TLI right here instead of doing it below per function. TLI is per module.

Reply:
Function may have extra attributes that determines whether to ignore certain builtin functions:
https://github.com/llvm/llvm-project/blob/da97d12cc05a2474cd893b996c0d789f766d97b8/llvm/include/llvm/Analysis/TargetLibraryInfo.h#L225